streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java

```diff
@@ -0,0 +1,399 @@
+                    batch.delete(oldColumnFamily, record.key);
+                    batch.put(newColumnFamily, record.key, record.value);
+                }
+            }
+        }
+
+        @Override
+        public void close() {
+            oldColumnFamily.close();
+            newColumnFamily.close();
+        }
+
+        @Override
+        public void toggleDbForBulkLoading() {
+            try {
+                db.compactRange(oldColumnFamily, true, 1, 0);
+            } catch (final RocksDBException e) {
+                throw new ProcessorStateException("Error while range compacting during restoring  store " + name, e);
+            }
+            try {
+                db.compactRange(newColumnFamily, true, 1, 0);
+            } catch (final RocksDBException e) {
+                throw new ProcessorStateException("Error while range compacting during restoring  store " + name, e);
+            }
+        }
+    }
+
+    private class RocksDBDualCFIterator extends AbstractIterator<KeyValue<Bytes, byte[]>>
+        implements KeyValueIterator<Bytes, byte[]> {
+
+        // RocksDB's JNI interface does not expose getters/setters that allow the
+        // comparator to be pluggable, and the default is lexicographic, so it's
+        // safe to just force lexicographic comparator here for now.
+        private final Comparator<byte[]> comparator = Bytes.BYTES_LEXICO_COMPARATOR;
+
+        private final String storeName;
+        private final RocksIterator iterWithTimestamp;
+        private final RocksIterator iterNoTimestamp;
+
+        private volatile boolean open = true;
+
+        private byte[] nextWithTimestamp;
+        private byte[] nextNoTimestamp;
+        private KeyValue<Bytes, byte[]> next;
+
+        RocksDBDualCFIterator(final String storeName,
+                              final RocksIterator iterWithTimestamp,
+                              final RocksIterator iterNoTimestamp) {
+            this.iterWithTimestamp = iterWithTimestamp;
+            this.iterNoTimestamp = iterNoTimestamp;
+            this.storeName = storeName;
+        }
+
+        @Override
+        public synchronized boolean hasNext() {
+            if (!open) {
+                throw new InvalidStateStoreException(String.format("RocksDB iterator for store %s has closed", storeName));
+            }
+            return super.hasNext();
+        }
+
+        @Override
+        public synchronized KeyValue<Bytes, byte[]> next() {
+            return super.next();
+        }
+
+        @Override
+        public KeyValue<Bytes, byte[]> makeNext() {
+            if (nextNoTimestamp == null && iterNoTimestamp.isValid()) {
+                nextNoTimestamp = iterNoTimestamp.key();
+            }
+
+            if (nextWithTimestamp == null && iterWithTimestamp.isValid()) {
+                nextWithTimestamp = iterWithTimestamp.key();
+            }
+
+            if (nextNoTimestamp == null) {
+                if (nextWithTimestamp == null) {
```

Comment:
should both iterators also be reporting `!isValid` here as well? I'm finding he rocksdb iterator api a little confusing...

I guess if we never allow a null key into the store, then this is an effective way to check for the end of the iteration.

Reply:
`null` keys are not allowed -- that is the assumption in this check. We can add the `!isValid` as safe guard if you want.

Reply:
I don't feel strongly about it. If we enforce the "no null keys" invariant, then they are equivalent.

It seems mildly confusing that we essentially have two different methods of determining when the iterator has run out of data. I leave it up to you.

Reply:
\cc @guozhangwang WDYT?

Reply:
Ack. Updating this.